src/optimizer/asp_optimizer.cpp

```diff
@@ -25,54 +25,67 @@ void ASPOptimizer::visitOperator(planner::LogicalOperator* op) {
 
 void ASPOptimizer::visitHashJoin(planner::LogicalOperator* op) {
     auto hashJoin = (LogicalHashJoin*)op;
-    if (!canApplyASP(op)) {
+    if (hashJoin->getChild(0)->getOperatorType() == LogicalOperatorType::ACCUMULATE) {
+        // No ASP if probe side has already been accumulated. This can be solved.
+        return;
+    }
+    auto probeSideFilterCollector = LogicalFilterCollector();
+    probeSideFilterCollector.collect(op->getChild(0).get());
+    if (!probeSideFilterCollector.hasOperators()) {
+        // Probe side is not selective so we don't apply ASP.
+        return;
+    }
+    auto scanNodes = resolveScanNodesToApplySemiMask(op);
+    if (scanNodes.empty()) {
         return;
     }
-    auto joinNodeID = hashJoin->getJoinNodeIDs()[0];
     // apply ASP
     hashJoin->setInfoPassing(planner::HashJoinSideWayInfoPassing::LEFT_TO_RIGHT);
-    auto semiMasker = std::make_shared<LogicalSemiMasker>(joinNodeID, op->getChild(0));
-    semiMasker->computeFlatSchema();
-    auto accumulate = std::make_shared<LogicalAccumulate>(std::move(semiMasker));
+    auto currentChild = op->getChild(0);
```

Comment:
I think it's more clear to write as `    auto currentChild = hashJoin->getChild(0);` here.